burstphoto/align.swift

```diff
@@ -490,190 +1135,201 @@ func robust_merge(_ ref_texture: MTLTexture, _ ref_texture_blurred: MTLTexture,
+    command_encoder.setBytes([Int32(tile_info.tile_size_merge)], length: MemoryLayout<Int32>.stride, index: 0)
+    command_encoder.dispatchThreads(threads_per_grid, threadsPerThreadgroup: threads_per_thread_group)
+    command_encoder.endEncoding()
+    command_buffer.commit()
     
-    // for Fuji X-Trans, the Bayer patter has size 6x6 but for conversion to grayscale, we can pretend like it's just 3x3
-    // we assume that all 6x6 patterns behave this way - is that correct?
-    if (mosaic_pettern_width == 6) {mosaic_pettern_width = 3}
+    return rms_texture
+}
+
+
+func forward_ft(_ in_texture: MTLTexture, _ out_texture_ft: MTLTexture, _ tmp_texture_ft: MTLTexture, _ tile_info: TileInfo, mode: String) {
+  
+    var cosine_factor = 0.500    
+    // adapt cosine factor dependent on tile size
+    if (Int32(tile_info.tile_size_merge) == 8) {cosine_factor = 0.530}
+    else if (Int32(tile_info.tile_size_merge) == 16) {cosine_factor = 0.505}
     
-    // set alignment params
-    let min_image_dim = min(ref_texture.width, ref_texture.height)
-    var downscale_factor_array = [mosaic_pettern_width]
-    var search_dist_array = [2]
-    var tile_size_array = [tile_size]
-    var res = min_image_dim / downscale_factor_array[0]
-    while (res > max_min_layer_res) {
-        downscale_factor_array.append(2)
-        search_dist_array.append(2)
-        tile_size_array.append(max(tile_size_array.last!/2, 8))
-        res /= 2
-    }
+    let command_buffer = command_queue.makeCommandBuffer()!
+    let command_encoder = command_buffer.makeComputeCommandEncoder()!
+    // either use discrete Fourier transform or highly-optimized fast Fourier transform
+    let state = (mode == "DFT" ? forward_dft_state : forward_fft_state)
+    command_encoder.setComputePipelineState(state)
+    let threads_per_grid = MTLSize(width: tile_info.n_tiles_x, height: tile_info.n_tiles_y, depth: 1)
+    let threads_per_thread_group = get_threads_per_thread_group(state, threads_per_grid)
+    command_encoder.setTexture(in_texture, index: 0)
+    command_encoder.setTexture(tmp_texture_ft, index: 1)
+    command_encoder.setTexture(out_texture_ft, index: 2)
+    command_encoder.setBytes([Int32(tile_info.tile_size_merge)], length: MemoryLayout<Int32>.stride, index: 0)
+    command_encoder.setBytes([Float32(cosine_factor)], length: MemoryLayout<Float32>.stride, index: 1)
+    command_encoder.dispatchThreads(threads_per_grid, threadsPerThreadgroup: threads_per_thread_group)
+    command_encoder.endEncoding()
+    command_buffer.commit()
+}
+
+
+func backward_ft(_ in_texture_ft: MTLTexture, _ tmp_texture_ft: MTLTexture, _ tile_info: TileInfo, _ n_textures: Int, mode: String) -> MTLTexture {
     
-    // build reference pyramid
-    let ref_pyramid = build_pyramid(ref_texture, downscale_factor_array)
+    let out_texture_descriptor = MTLTextureDescriptor.texture2DDescriptor(pixelFormat: .rgba32Float, width: in_texture_ft.width/2, height: in_texture_ft.height, mipmapped: false)
```

Comment:
Why is this width/2 but full height?

Reply:
The textures used for the Fourier transformed image (..._ft) store both the real and the imaginary part in the same texture. As a consequence, they are larger by a factor of two in one dimension compared to the image in the image space.